Math.md

```diff
@@ -0,0 +1,96 @@
+ let b = 12
+ let n = b / a   (3)
+ let r = b % a   (0)
+ let product = a * b   (48)
+ let sum = a + b   (16)
+ let difference = b - a   (8)
+ ````
+
+ and using Floats...  (because there's a decimal the compiler knows it's a Float)
+
+````Swift
+ let x = 5.5
+ let y = 10.0
+ let n = y / x  (1.8181...)
+ let r = y % x   (4.5)
+ let proudct = x * y   (55)
+ let sum = x + y  (15.5)
+ let difference = y - x  (4.5)
+ ````
+ 
+ So far this should all remind you of your days back in math class with Mr. Mcgregor.  There is, however, something that you may find surprising about how Swift handles math.
+ 
+ What if we multiply an Int and Float, like a * x as shown above?  Well, the compiler gives us an error which seems very unfriendly!  Why is that?
+ 
+Remember that one of the key features of Swift is that it's a type-safe language; that means its very, very particular about what kind of types it expects in any given situation and will unceremoniously give you an error whenever you provide it a type it doesn't like.
+
+In Mathematical operations Swift only lets you use values of the same type because (as we've learned over the past several decades with langauges which "helpfully" convert and combine types for us) mixing types can lead to unexpected results such as loss of precision.
+
+If you want to mix types in mathematical expressions you must tell Swift explicitly what you want to convert.
+ 
+````Swift
+ let sum = a + Int(x)    (9)
+ let sum = Double(a) + x   (9.5)
+ ````
+ 
+ You can easily explore and experiment with mathetimatical operations in a Playground.
+ 
+ 
+##Functions
+ Functions, as mentioned in an earlier unit, are the primary means of reducing complexity and making programs easier to read.  Instead of having hundreds or thousands (or tens of thousands) of lines of code, you can break that up in functions to make you code easier to understand.  Some functions are just a single line and others are hundreds of lines of code but a good rule of thumb is that a function should be between 10 and 100 lines of code; fewer lines of code than that and creating the function may not be worth the trouble, and more lines of code than that means you probably have more complexity than you want and should probably break that function into two or more functions if possible.
+ 
+ Most functions return some kind of value; either something which is retrieved, computed, or just a code indicating success or failure.  For example, here's a function which returns the average of two Doubles:
+
+````Swift
+ func average(a:Double, b:Double, c:Double) -> Double {
+ 	return ( a + b + c ) / 3.0
+ 	}
+ 	````
+ 	
+ You can also have a function which returns the average of three Ints:
+
+````Swift
+ 	 func average(a:Double, b:Double, c:Double) -> Double {
+ 	return ( a + b + c ) / 3.0
+ 	}
+ 	````
+ 	
+It's also perfectly OK to have a function which doesn't take any arguments or return any value:
+ 
+````Swift
+ 	func sayHello() {
+ 		print("Hello")
+ 	}
+````
```

Comment:
I love showing these kinds of examples here like you did. I think we can be slightly more explicit in what we show/explain to the student here, like so:

```swift
func average(a: Int, _ b: Int) -> Int {
    
    return (a + b) / 2
    
}

let midTermGrade = 70
let finalGrade = 100

average(midTermGrade, finalGrade) // 85
```
Here we're creating a function named average which takes in two arguments of type `Int` and returns back an `Int`. The implementation of this function takes the sum of the two arguments being passed in, divides that sum by 2 and returns back the result which will be of type `Int`. Here, we're calling on this function by passing in our mid term and final grades to get back the average.

-----

```swift
func average(a: Double, _ b: Double, _ c: Double) -> Double {
    
    return (a + b + c) / 3.0
    
}

let first = 42.10
let second = 44.98
let third = 50.31

average(first, second, third) // 45.79666666666666
```

Here we're creating a function named `average` which takes in three arguments, all of type `Double` and returns back a `Double`. The implementation of this function takes the sum of the three arguments then divides that sum by 3.0. It returns that value of type `Double` to the caller of this function. Here we provide an example of calling on this function passing in three arguments of type `Double`.

-----

It's also perfectly OK to have a function which doesn't take any arguments or return any value:

```swift
func sayHello() {
    print("Hello")
}

sayHello()
// prints "Hello"
```